core/src/main/java/hudson/security/SecurityRealm.java

```diff
@@ -484,6 +485,44 @@ public Filter createFilter(FilterConfig filterConfig) {
      */
     public static final SecurityRealm NO_AUTHENTICATION = new None();
 
+    /**
+     * Perform a calculation where we should go back after sucessfull login
+     *
+     * @return Encoded URI where we should go back after sucessfull login or "/" if no way back
+     *
+     * @since TODO
+     */
+    public static String getFrom() {
+        String from = null;
+        final StaplerRequest request = Stapler.getCurrentRequest();
+
+        if (request.getSession(false) != null) {
+            from = (String) request.getSession().getAttribute("from");
+        }
+
+        if (from == null) {
+            from = request.getParameter("from");
+        }
+
+        final String requestURI = request.getRequestURI();
+        if (from == null && requestURI != null
+                && requestURI.compareTo("/loginError") != 0 && requestURI.compareTo("/login") != 0) {
+            from = requestURI;
+        }
+
+        if (from == null || from.trim().isEmpty()) {
+            from = "/";
+        }
+
+        try {
+            from = java.net.URLEncoder.encode(from, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            from = "/";
```

Comment:
Sure, but if it is occurred for whatever reason I use this as a safety and reasonable (IMO) fallback. Do you prefer to raise an exception in that case?